test/tools/runner/hooks/leak_checker.ts

```diff
@@ -0,0 +1,182 @@
+    this.clients = new Set<MongoClient>();
+    this.sessionPools = new Set<ServerSessionPool>();
+  }
+
+  setupSessionLeakChecker() {
+    const leakChecker = this;
+    ServerSessionPool.prototype.acquire = function (...args) {
+      leakChecker.sessionPools.add(this);
+
+      this[LeakChecker.kAcquiredCount] ??= 0;
+      this[LeakChecker.kAcquiredCount] += 1;
+
+      return LeakChecker.originalAcquire.call(this, ...args);
+    };
+
+    ServerSessionPool.prototype.release = function (...args) {
+      if (!(LeakChecker.kAcquiredCount in this)) {
+        throw new Error('releasing before acquiring even once??');
+      } else {
+        this[LeakChecker.kAcquiredCount] -= 1;
+      }
+
+      return LeakChecker.originalRelease.call(this, ...args);
+    };
+  }
+
+  setupClientLeakChecker() {
+    const leakChecker = this;
+    MongoClient.prototype.connect = function (...args) {
+      leakChecker.clients.add(this);
+      this[LeakChecker.kConnectCount] ??= 0;
+
+      const lastArg = args[args.length - 1];
+      const lastArgIsCallback = typeof lastArg === 'function';
+      if (lastArgIsCallback) {
+        const argsWithoutCallback = args.slice(0, args.length - 1);
+        return LeakChecker.originalConnect.call(this, ...argsWithoutCallback, (error, client) => {
+          if (error == null) {
+            this[LeakChecker.kConnectCount] += 1; // only increment on successful connects
+          }
+          return lastArg(error, client);
+        });
+      } else {
+        return LeakChecker.originalConnect.call(this, ...args).then(client => {
+          this[LeakChecker.kConnectCount] += 1; // only increment on successful connects
+          return client;
+        });
+      }
+    };
+
+    MongoClient.prototype.close = function (...args) {
+      this[LeakChecker.kConnectCount] ??= 0; // prevents NaN, its fine to call close on a client that never called connect
+      this[LeakChecker.kConnectCount] -= 1;
+      return LeakChecker.originalClose.call(this, ...args);
+    };
+  }
+
+  setup() {
+    this.setupSessionLeakChecker();
+    this.setupClientLeakChecker();
+  }
+
+  reset() {
+    for (const sessionPool of this.sessionPools) {
+      delete sessionPool[LeakChecker.kAcquiredCount];
+    }
+    ServerSessionPool.prototype.acquire = LeakChecker.originalAcquire;
+    ServerSessionPool.prototype.release = LeakChecker.originalRelease;
+    this.sessionPools.clear();
+
+    for (const client of this.clients) {
+      delete client[LeakChecker.kConnectCount];
+    }
+    MongoClient.prototype.connect = LeakChecker.originalConnect;
+    MongoClient.prototype.close = LeakChecker.originalClose;
+    this.clients.clear();
+  }
+
+  assert() {
+    for (const pool of this.sessionPools) {
+      expect(pool[LeakChecker.kAcquiredCount], 'ServerSessionPool acquired count').to.equal(0);
+    }
+    for (const client of this.clients) {
+      expect(client[LeakChecker.kConnectCount], 'MongoClient connect count').to.be.lessThanOrEqual(
```

Comment:
why can't we check strict equality? doesn't < 0 imply that something went wrong?

Reply:
We don't do it often but we have at least that one test that closes an unconnected client, so the count will go negative. And I'm thinking like there could be tests that close the shared client from the describe block before the afterEach gets to it and there's no harm in that. close on a closed client is a no-op

Reply:
alright